File: background/python/algo/leetcode/question1.py
"""
给定一个整数数组 nums 和一个整数目标值 target，请你在该数组中找出 和为目标值 target  的那 两个 整数，并返回它们的数组下标。

你可以假设每种输入只会对应一个答案，并且你不能使用两次相同的元素。

你可以按任意顺序返回答案。
"""

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_dict: Dict[int, int] = {}
        for i, num in enumerate(nums):
            nextNum = target - num
            if nextNum in nums_dict:
                return [i, nums_dict[nextNum]]
            nums_dict[num] = i
        return []

File: background/leetcode/Question1.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Question1 {
    static class Solution {
        public int[] twoSum(int[] nums, int target) {
            Map<Integer, Integer> map = new HashMap<>();
            for (int i = 0; i < nums.length; i++) {
                int complement = target - nums[i];
                if (map.containsKey(complement)) {
                    return new int[]{map.get(complement), i};
                }
                map.put(nums[i], i);
            }
            return null;
        }
    }

    public static void run() {
        Solution s = new Solution();

        int[][] testCases = {
            {2, 7, 11, 15},
            {3, 2, 4},
            {3, 3}
        };
        int[] targets = {9, 6, 6};
        int[][] expected = {
            {0, 1},
            {1, 2},
            {0, 1}
        };

        for (int i = 0; i < testCases.length; i++) {
            int[] result = s.twoSum(testCases[i], targets[i]);
            boolean pass = Arrays.equals(result, expected[i]);
            System.out.println("Test " + (i + 1) + ": " + (pass ? "PASS" : "FAIL")
                    + " " + Arrays.toString(result));
        }
    }

    public static void main(String[] args) {
        run();
    }
}